docs/designs/supporting_pytorch.md

```diff
@@ -0,0 +1,358 @@
+            # train your data...
+            print('Epoch: ', epoch, '| Step: ', step, '| batch x: ',
+            batch_x, '| batch y: ', batch_y)
+            # print('Epoch: ', epoch, '| Step: ', step, '| batch x: ',
+            #       batch_x.numpy(), '| batch y: ', batch_y.numpy())
+if __name__ == '__main__':
+    show()
+```
+
+### MNIST Example
+
+This dataset is in numpy array format, and has been stored using pickle, a
+python-specific format for serializing data. Each image is 28 x 28, and is being
+ stored as a flattened row of length 784 (=28x28). Let’s take a look at one; we
+ need to reshape it to 2d first.
+
+```python
+plt.imshow(train_data.train_data[0].numpy(), cmap='gray')
+plt.title('%i' % train_data.train_labels[0])
+plt.show()
+```
+
+PyTorch uses `torch.tensor`, rather than `numpy` arrays.
+
+```python
+x_train, y_train, x_valid, y_valid = map(
+    torch.tensor, (x_train, y_train, x_valid, y_valid)
+)
+```
+
+Or we can convert our data like this.
+
+```python
+train_data = torchvision.datasets.MNIST(
+    root='./mnist/',
+    train=True,                                     # this is training data
+    transform=torchvision.transforms.ToTensor(),    # Converts a PIL.Image or numpy.ndarray to
+                                                    # torch.FloatTensor of shape (C x H x W) and normalize in the range [0.0, 1.0]
+    download=DOWNLOAD_MNIST,
+)
+```
+
+## Training Loop
+
+ElasticDL holds master-worker architecture. Workers get model parameters from
+parameter server (PS), compute gradients using different training data, and send
+computed gradients to PS. PS iteratively updates these model parameters using
+gradients sent by workers. In our training loop, the gradient information of the
+neural network needs to be calculated and transmitted between the ps and the worker.
+The advanced API in PyTorch such as `torch.optim` is not available,we had to
+update the value of each parameter by name, and manually zero the gradient of
+each parameter, as follows:
+
+```python
+with torch.no_grad():
+    weights -= weights.grad * lr
+    bias -= bias.grad * lr
+    weights.grad.zero_()
+    bias.grad.zero_()
+```
+
+`torch.no_grad()` context is necessary because we don't want to record these
+operations in the next gradient calculation.
+To go further, we can use `model.parameters()` and `model.zero_grad()` (defined
+by PyTorch for `nn.Module`) to make these steps more concise, and there will be
+no errors of forgetting some parameters, especially when we build a complex model:
+
+```python
+with torch.no_grad():
+    for param in model.parameters(): param -= param.grad * lr
+    model.zero_grad()
+```
+
+## Work with PS_client
+
+This [document](https://github.com/sql-machine-learning/elasticdl/blob/develop/docs/designs/parameter_server.md)
+describes the design of a distributed parameter server for ElasticDL.
+
+```python
+class PSClient(object):
```

Comment:
We could give a reference for the `PSClient` code, but focus on describing how to use the API.